### ISSUE TITLE
Server crashes if transmitter and receiver are connected by a digiline

### ISSUE THREAD

**Issue by numberZero** [lordcirth/minetest-wireless#1]
As soon as any message is sent to the transmitter, it is repeated by the receiver; if they are connected, that becomes a loop that causes stack overflow (due to synchronous processing) after transmitting the message 1.5k times.

See also minetest-mods/mesecons#468
@neoh4x0r Can’t reproduce without transmitter and receiver being connected by wire, LuaC overheats after processing 30 message copies.

**Comment by lordcirth** [comment]
Well, it's cool that people are still using this! I haven't touched Minetest or Lua in years, so this might be hard for me to fix. HybridDog took over minetest-digipad a while ago, and I see they are still active. Perhaps they are interested in taking over? If not, I could take a look later.

**Comment by neoh4x0r** [comment]
EDIT: A solution to the problem is listed at the bottom
-----
The layout for this is as follow (only this time, with such a small world no stack overflow was generated, but one lua controller did burn up -- I would imagine with a larger world and more going on that a stackoverflow would be generated before the controller burnt up).

You can view the the layout [here](https://www.dropbox.com/s/4qzz54od3qs9v7g/layout.jpg?dl=0)
And a video of it can bee seen [here](https://www.dropbox.com/s/fbb0wleyr793xp5/minetest-luacontroller-2019-07-14_17.24.10.mp4?dl=0) (I apologize for the video quality)

In the layout the various luacontrollers are labeled: lua1, lua2, and lua_lcd.

These are the lua scripts I am using for each controller:
-------------
lua1.lua:
```
if event.type == "digiline" then
 if event.channel == "lua1_b" then
  port.b = not port.b
 end
end

if pin.d then
 digiline_send("lcd","luac_#1")
 digiline_send("lua2_d","on") 
end
```

lua2.lua:
```
if pin.b then
 digiline_send("lcd","luac_#2")
 digiline_send("lua1_b","on")
end
```

lua_lcd.lua:
```
if event.type == "digiline" then
 if event.channel == "lcd" then
    port.b = not port.b
 end
 if event.channel == "lua1_b" then
   port.c = not port.c
 end
 if event.channel == "lua2_d" then
   port.d = not port.d
 end
end
```

When turning the switch on/off on lua2 (on port b): (**works perfectly**)
- toggles port b and c  on lua_lcd 
- toggles port b on lua1
- sets lcd text to "luac_#2"


When turning the switch on/off on lua1 (on port d): (**this is the problem controller**)
- toggles port b and d  on lua_lcd 
- toggles port d on lua2 (not implemented, because lua2 does not have a receiver)
- sets lcd text to "luac_#1"

The problem is that lua1 overheats just after turning the switch on port d.
For an instant port b on lua_lcd turns on (and then off), while port d on lua_lcd toggles as expected.

It seems that the luacontroller cannot call digiline_send (transmit) and subscribe to a digiline event (receive) at the same time.

The intention is to have lua1 be able to both send and receive digilines messages (wirelessly).

I'll assume that when lua1 transmits (connected to port a) the receiver actives (on port c) and causes the lua controller to execute its script again (since the switch on port d is still active) it creates a endless loop (hence stack overflow).

honestly the code for lua1 probably needs to be this 
(**barring a more elegant solution, I think this would fix the issue**)
lua1.lua:
```
if event.type == "digiline" then
 if event.channel == "lua1_b" then
  port.b = not port.b
 end
return -- block re-execution of script
end

if pin.d then
 digiline_send("lcd","luac_#1")
 digiline_send("lua2_d","on") 
end
```

Testing the above change to lua1.lua -- it behaves without burning the controller or generating any other errors.

In other words calling digiline_send such that a digiline_event is recieved by the same controller (even if it is not destined for that controller) causes the script to be re-executed and an endless loop generated.

**Comment by numberZero** [comment]
> I'll assume that when lua1 transmits (connected to port a) the receiver actives (on port c) and causes the lua controller to execute its script again (since the switch on port d is still active) it creates a endless loop

Exactly. And that burns the controller.

> (hence stack overflow).

No. I’m almost sure you’ve connected the transmitter and the receiver accidentally in your original setup. That will burn the server ofc as there is no protection at all. I’ll write a fix soon.

> In other words calling digiline_send such that a digiline_event is recieved by the same controller (even if it is not destined for that controller) causes the script to be re-executed

Of course, as there is no other way to find out whether that message is destined for that controller.

> and an endless loop generated.

Only if the controller repeats transmission.

Note that you can build such loop without wireless transmission, using a pair of LuaCs instead. Like connecting lua1 and lua2 together and enabling both.

@lordcirth Nvm, I can take care of this mod myself.
